File: tests/fixtures/open_redirect/go/safe_host_allowlist_redirect.go
package handler

import (
	"net/http"
	"net/url"
)

const allowedHost = "trusted.example.com"

// Safe: tainted query value parsed via `url.Parse` then host pinned against
// `allowedHost`.  Multi-statement form — `parsed, err := url.Parse(target)`
// happens on a separate line from the `parsed.Host == allowedHost` check.
// Recognised by PredicateKind::HostAllowlistValidated which clears
// Cap::OPEN_REDIRECT on the validated branch.
func SafeHostAllowlist(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("next")
	parsed, err := url.Parse(target)
	if err != nil {
		http.Redirect(w, r, "/", http.StatusFound)
		return
	}
	if parsed.Host == allowedHost {
		http.Redirect(w, r, parsed.String(), http.StatusFound)
		return
	}
	http.Redirect(w, r, "/", http.StatusFound)
}

File: tests/fixtures/open_redirect/rust/safe_host_allowlist_redirect.rs
// Safe: tainted env value parsed via `url::Url::parse` then host pinned
// against `ALLOWED_HOST`.  Multi-statement form — `parsed = Url::parse(x)`
// happens on a separate line from the `parsed.host_str() == Some(ALLOWED)`
// check.  Recognised by PredicateKind::HostAllowlistValidated which clears
// Cap::OPEN_REDIRECT on the validated branch.
use axum::response::Redirect;
use std::env;
use url::Url;

const ALLOWED_HOST: &str = "trusted.example.com";

fn bounce() -> Redirect {
    let next = env::var("NEXT").unwrap_or_default();
    let parsed = Url::parse(&next).unwrap();
    if parsed.host_str() == Some(ALLOWED_HOST) {
        return Redirect::to(parsed.as_str());
    }
    Redirect::permanent("/")
}
